server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -168,8 +186,64 @@ private void doRedirectFailures() {
+            // Get new cluster state that includes any potential failure store rollovers.
+            var rolledOverState = observer.setAndGetObservedState();
+            Map<ShardId, List<BulkItemRequest>> requestsByShard = drainAndGroupRedirectsByShards(rolledOverState);
+            executeBulkRequestsByShard(requestsByShard, rolledOverState, this::completeBulkOperation);
+        };
+        rollOverFailureStores(clusterState, executeRedirectRequests);
+    }
+
+    /**
+     * Send rollover requests for all failure stores that need it. After all requests have completed, we execute the given runnable.
+     * Any failures while rolling over will be added to the {@link BulkItemResponse} entries of the index requests that were redirected to
+     * the failure store that failed to roll over.
+     */
+    private void rollOverFailureStores(ClusterState clusterState, Runnable runnable) {
+        // Skip allocation of some objects if we don't need to roll over anything.
+        if (failureStoresToBeRolledOver.isEmpty()
+            || featureService.clusterHasFeature(clusterState, LazyRolloverAction.FAILURE_STORE_LAZY_ROLLOVER) == false) {
+            runnable.run();
+            return;
+        }
+        try (RefCountingRunnable refs = new RefCountingRunnable(runnable)) {
+            for (String dataStream : failureStoresToBeRolledOver) {
+                RolloverRequest rolloverRequest = new RolloverRequest(dataStream, null);
+                rolloverRequest.setIndicesOptions(
+                    IndicesOptions.builder(rolloverRequest.indicesOptions())
+                        .failureStoreOptions(new IndicesOptions.FailureStoreOptions(false, true))
+                        .build()
+                );
+                // We are executing a lazy rollover because it is an action specialised for this situation, when we want an
+                // unconditional and performant rollover.
+                rolloverClient.execute(LazyRolloverAction.INSTANCE, rolloverRequest, ActionListener.releaseAfter(new ActionListener<>() {
+
+                    @Override
+                    public void onResponse(RolloverResponse result) {
+                        // A successful response has rolled_over false when in the following cases:
+                        // - A request had the parameter lazy or dry_run enabled
+                        // - A request had conditions that were not met
+                        // Since none of the above apply, getting a response with rolled_over false is considered a bug
+                        // that should be caught here and inform the developer.
+                        assert result.isRolledOver() : "An successful lazy rollover should always result in a rolled over data stream";
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        for (BulkItemRequest failureStoreRedirect : failureStoreRedirects) {
```

Comment:
This loop makes me uneasy. In practice it's probably closer to O(n) where n is the number of documents in the bulk operation, but technically it is O(n^2) in the worst case, which would be if each document was destined to a different failure store and each rollover failed.

It looks like this is how things currently work in the bulk operation for lazy rollover, and I don't know how frequently we see lazy rollover failures, so you'd need to torture your cluster in a unique way to get this to run n^2, but it's possible. Maybe we should file a follow up to collect all the failed rollovers and traverse the documents one time using a set of failed rollovers to update the response. We'd need to fix it here and in the transport action.

Reply:
Filed https://github.com/elastic/elasticsearch/issues/108838